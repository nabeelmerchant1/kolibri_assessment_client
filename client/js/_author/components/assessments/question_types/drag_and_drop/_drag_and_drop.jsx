import React              from 'react';
import { connect }        from 'react-redux';
import * as AssetActions  from '../../../../../actions/qbank/assets';
import TargetArea         from './target_area';
import DragArea           from './drag_area';
import Feedback           from '../question_common/single_feedback';
import localize           from '../../../../locales/localize';

function select(state) {
  return {
    loadingMedia: state.media.loading,
    images: state.media.image,
  };
}

export class DragAndDrop extends React.Component {
  static propTypes = {
    item: React.PropTypes.shape({
      id: React.PropTypes.string,
      bankId: React.PropTypes.string,
      question: React.PropTypes.shape({}),
    }).isRequired,
    images: React.PropTypes.shape({}).isRequired,
    updateItem: React.PropTypes.func.isRequired,
    addMediaToQuestion: React.PropTypes.func.isRequired,
    updateChoice: React.PropTypes.func.isRequired,
    localizeStrings: React.PropTypes.func.isRequired,
    loadingMedia: React.PropTypes.bool,
    isActive: React.PropTypes.bool,
    language: React.PropTypes.string.isRequired,
  };

  constructor() {
    super();
    this.state = {};
  }

  uploadMedia(file, where, metadata, newMedia) {
    this.props.addMediaToQuestion(
      file,
      this.props.item.bankId,
      this.props.item.id,
      `question.${where}`,
      metadata,
      newMedia,
      this.props.language
    );
  }

  render() {
    const { question, id, bankId } = this.props.item;

    const strings = this.props.localizeStrings('dragAndDrop');
    return (
      <div style={{ display: this.props.isActive ? 'block' : 'none' }}>
        <TargetArea
          id={id}
          question={question}
          loadingMedia={this.props.loadingMedia}
          images={this.props.images}
          uploadMedia={(file, where, metadata, newMedia) =>
            this.uploadMedia(file, where, metadata, newMedia)}
          visibleZones={question.visibleZones}
          setVisible={visibleZones => this.props.updateItem({ question: { visibleZones } }, true)}
          editZone={(zoneId, attributes) =>
            this.props.updateChoice(id, zoneId, attributes, null, 'zones')}
          language={this.props.language}

        />
        <div className="au-c-drop-zone__answers__label">{strings.draggableAnswers}</div>
        <DragArea
          dropObjects={question.dropObjects}
          zones={question.zones}
          updateDropObject={(choiceId, choice, fileIds) =>
            this.props.updateChoice(id, choiceId, choice, fileIds, 'dropObjects')}
          loadingMedia={this.props.loadingMedia}
          images={this.props.images}
          uploadMedia={(file, where, metadata, newMedia) =>
            this.uploadMedia(file, where, metadata, newMedia)}
          language={this.props.language}
        />
        <div className="au-c-question__feedback">
          <Feedback
            language={this.props.language}
            updateItem={this.props.updateItem}
            feedbackType="correctFeedback"
            feedback={question.correctFeedback}
            labelText="Correct Feedback"
            bankId={bankId}
          />
          <Feedback
            language={this.props.language}
            updateItem={this.props.updateItem}
            feedbackType="incorrectFeedback"
            feedback={question.incorrectFeedback}
            labelText="Incorrect Feedback"
            bankId={bankId}
          />
        </div>
      </div>
    );
  }
}

export default connect(select, AssetActions)(localize(DragAndDrop));
