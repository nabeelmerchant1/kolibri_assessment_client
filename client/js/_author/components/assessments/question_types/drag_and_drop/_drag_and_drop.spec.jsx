import React            from 'react';
import { shallow }      from 'enzyme';
import { DragAndDrop }  from './_drag_and_drop';

jest.mock('../../../../../libs/assets.js');

describe('drag_and_drop component', () => {
  let props;
  let result;
  let uploadedMedia;

  beforeEach(() => {
    uploadedMedia = false;
    props = {
      item: {
        id: '',
        bankId: '',
        question: {
          id: '7',
          dropObjects: {},
          zones: {},
        },
      },
      images: {},
      updateItem: () => {},
      addMediaToQuestion: () => { uploadedMedia = true; },
      updateChoice: () => {},
      loadingMedia: false,
      language: 'eng',
    };
    result = shallow(<DragAndDrop {...props} />);
  });

  it('renders the component as a snapshot', () => {
    expect(result).toMatchSnapshot();
  });

  it('executes the upload media function', () => {
    expect(uploadedMedia).toBeFalsy();
    result.instance().uploadMedia();
    expect(uploadedMedia).toBeTruthy();
  });
});
