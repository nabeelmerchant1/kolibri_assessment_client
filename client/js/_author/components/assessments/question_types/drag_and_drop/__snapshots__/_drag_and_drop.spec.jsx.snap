// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drag_and_drop component renders the component as a snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div
    style={
        Object {
            "display": "none",
          }
    }
>
    <TargetArea
        editZone={[Function]}
        id=""
        images={Object {}}
        language="eng"
        loadingMedia={false}
        question={
            Object {
                "dropObjects": Object {},
                "id": "7",
                "zones": Object {},
              }
        }
        setVisible={[Function]}
        uploadMedia={[Function]}
        visibleZones={undefined}
    />
    <div
        className="au-c-drop-zone__answers__label"
    >
        Draggable answers
    </div>
    <DragArea
        dropObjects={Object {}}
        images={Object {}}
        language="eng"
        loadingMedia={false}
        localizeStrings={[Function]}
        updateDropObject={[Function]}
        uploadMedia={[Function]}
        zones={Object {}}
    />
    <div
        className="au-c-question__feedback"
    >
        <SingleFeedback
            bankId=""
            feedback={undefined}
            feedbackType="correctFeedback"
            labelText="Correct Feedback"
            language="eng"
            updateItem={[Function]}
        />
        <SingleFeedback
            bankId=""
            feedback={undefined}
            feedbackType="incorrectFeedback"
            labelText="Incorrect Feedback"
            language="eng"
            updateItem={[Function]}
        />
    </div>
</div>,
  "nodes": Array [
    <div
      style={
            Object {
                  "display": "none",
                }
      }
>
      <TargetArea
            editZone={[Function]}
            id=""
            images={Object {}}
            language="eng"
            loadingMedia={false}
            question={
                  Object {
                        "dropObjects": Object {},
                        "id": "7",
                        "zones": Object {},
                      }
            }
            setVisible={[Function]}
            uploadMedia={[Function]}
            visibleZones={undefined}
      />
      <div
            className="au-c-drop-zone__answers__label"
      >
            Draggable answers
      </div>
      <DragArea
            dropObjects={Object {}}
            images={Object {}}
            language="eng"
            loadingMedia={false}
            localizeStrings={[Function]}
            updateDropObject={[Function]}
            uploadMedia={[Function]}
            zones={Object {}}
      />
      <div
            className="au-c-question__feedback"
      >
            <SingleFeedback
                  bankId=""
                  feedback={undefined}
                  feedbackType="correctFeedback"
                  labelText="Correct Feedback"
                  language="eng"
                  updateItem={[Function]}
            />
            <SingleFeedback
                  bankId=""
                  feedback={undefined}
                  feedbackType="incorrectFeedback"
                  labelText="Incorrect Feedback"
                  language="eng"
                  updateItem={[Function]}
            />
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <DragAndDrop
        addMediaToQuestion={[Function]}
        images={Object {}}
        item={
                Object {
                        "bankId": "",
                        "id": "",
                        "question": Object {
                          "dropObjects": Object {},
                          "id": "7",
                          "zones": Object {},
                        },
                      }
        }
        language="eng"
        loadingMedia={false}
        localizeStrings={[Function]}
        updateChoice={[Function]}
        updateItem={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": DragAndDrop {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "addMediaToQuestion": [Function],
          "images": Object {},
          "item": Object {
            "bankId": "",
            "id": "",
            "question": Object {
              "dropObjects": Object {},
              "id": "7",
              "zones": Object {},
            },
          },
          "language": "eng",
          "loadingMedia": false,
          "localizeStrings": [Function],
          "updateChoice": [Function],
          "updateItem": [Function],
        },
        "refs": Object {},
        "state": Object {},
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div
          style={
                    Object {
                              "display": "none",
                            }
          }
>
          <TargetArea
                    editZone={[Function]}
                    id=""
                    images={Object {}}
                    language="eng"
                    loadingMedia={false}
                    question={
                              Object {
                                        "dropObjects": Object {},
                                        "id": "7",
                                        "zones": Object {},
                                      }
                    }
                    setVisible={[Function]}
                    uploadMedia={[Function]}
                    visibleZones={undefined}
          />
          <div
                    className="au-c-drop-zone__answers__label"
          >
                    Draggable answers
          </div>
          <DragArea
                    dropObjects={Object {}}
                    images={Object {}}
                    language="eng"
                    loadingMedia={false}
                    localizeStrings={[Function]}
                    updateDropObject={[Function]}
                    uploadMedia={[Function]}
                    zones={Object {}}
          />
          <div
                    className="au-c-question__feedback"
          >
                    <SingleFeedback
                              bankId=""
                              feedback={undefined}
                              feedbackType="correctFeedback"
                              labelText="Correct Feedback"
                              language="eng"
                              updateItem={[Function]}
                    />
                    <SingleFeedback
                              bankId=""
                              feedback={undefined}
                              feedbackType="incorrectFeedback"
                              labelText="Incorrect Feedback"
                              language="eng"
                              updateItem={[Function]}
                    />
          </div>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          style={
                    Object {
                              "display": "none",
                            }
          }
>
          <TargetArea
                    editZone={[Function]}
                    id=""
                    images={Object {}}
                    language="eng"
                    loadingMedia={false}
                    question={
                              Object {
                                        "dropObjects": Object {},
                                        "id": "7",
                                        "zones": Object {},
                                      }
                    }
                    setVisible={[Function]}
                    uploadMedia={[Function]}
                    visibleZones={undefined}
          />
          <div
                    className="au-c-drop-zone__answers__label"
          >
                    Draggable answers
          </div>
          <DragArea
                    dropObjects={Object {}}
                    images={Object {}}
                    language="eng"
                    loadingMedia={false}
                    localizeStrings={[Function]}
                    updateDropObject={[Function]}
                    uploadMedia={[Function]}
                    zones={Object {}}
          />
          <div
                    className="au-c-question__feedback"
          >
                    <SingleFeedback
                              bankId=""
                              feedback={undefined}
                              feedbackType="correctFeedback"
                              labelText="Correct Feedback"
                              language="eng"
                              updateItem={[Function]}
                    />
                    <SingleFeedback
                              bankId=""
                              feedback={undefined}
                              feedbackType="incorrectFeedback"
                              labelText="Incorrect Feedback"
                              language="eng"
                              updateItem={[Function]}
                    />
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <DragAndDrop
    addMediaToQuestion={[Function]}
    images={Object {}}
    item={
        Object {
            "bankId": "",
            "id": "",
            "question": Object {
              "dropObjects": Object {},
              "id": "7",
              "zones": Object {},
            },
          }
    }
    language="eng"
    loadingMedia={false}
    localizeStrings={[Function]}
    updateChoice={[Function]}
    updateItem={[Function]}
/>,
}
`;
